### DISCUSSION TITLE
plugins/tableeditor maintain selection on target row

### DISCUSSION THREAD

**Discussion by kaizennet** [zim-desktop-wiki/zim-desktop-wiki#2269] [Ideas]
zim is the main notebook application on my Debian desktop.  One of the plugins I utilize most heavily is the Table Editor. The main inconvenience I was having with this plugin was when I selected a row to move it up or down the cursor selection would remain at the same row position instead of moving with the moved row.  

For example: If I had a task list of 20 items and wanted to move the item at row position 2 to the bottom of the list I would have to click the item, press down, click the item again at it's new position (now a row down), click down, etc.  18 times in total.

I'm inexperienced with Python but I forked the code and found the on_move_row function at line 687 in /zim/plugins/tableeditor.py

Adding the following lines of code at line 711 - the end of on_move_row before the on_open_link function (where are the curly brackets and semicolons?! the horror! ha) fixed the issue perfectly for me.  Now I can move rows up and down with ease.

           `if direction > 0:
			nextiter = model.iter_next(treeiter)
		else:
			nextiter = model.iter_previous(treeiter)
		path = model.get_path(nextiter)
		self.treeview.set_cursor(path, None, True)`

I'm unsure if the original behavior is by design so didn't want to try to check my code in before posting here.  Small change that made my life easier.  I'm working next on additional options to format text within the Table, 

Thank you

George

**Comment by introt** [comment]
Nice work! If the original behavior is by design, it's frankly bad design. Feel free to open a PR.

>  I'm working next on additional options to format text within the Table

Please see #818 (Refactor table editor to use nested textview instead of treeview) - we're unlikely to accept other implementations of text formatting inside tables, and using a nested textview would also allow images inside tables.

**Comment by kaizennet** [comment]
Thanks for the tip. I've been getting familiar with TextView functionality.  Hopefully I open the PR in the correct manner.  Is it recommended to only push through one isolated change at a time?

**Reply by introt** [reply]
Usually yes, as small singular patches are easier to review and are less likely to conflict with other's work. However, since you're basically rewriting the whole plugin, it's okay to just open a single PR for the whole thing as changes are mostly isolated to the plugin. If you need changes to APIs on the Zim side, those should preferably be submitted in advance. Feel free to open any number of draft PRs to ask questions etc!

**Comment by tridian-tn** [comment]
@kaizennet Hello! I'm looking at the various PRs and issues for the table plugin, and this idea/bug looks pretty essential. Did you ever get time to progress this? If not, no worries! Would you mind if I included your fix (attributed to you) in my PR?

**Reply by tridian-tn** [reply]
@kaizennet I've started a PR with a commit for this in there, just to get things started on other bits and bobs as well, but I'm happy to take that commit out if you'd prefer. Cheers!

**Reply by kaizennet** [reply]
Hi @tridian-tn .  Thanks for the message.  I need to get into the code again to see where I left off.   I remember I was able to make the tableeditor edit good enough for my usage  but when i tried to get into rewrite it as a textview plugin things got very complex in terms of trying to tie it into the existing code base.  At some point i just stepped away from it.

**Reply by tridian-tn** [reply]
@kaizennet No problem. I completely get that: I had a quick look at the idea of the textview refactor and it looked like far more work than I had time for (or tbh expertise with Python/GTK for). If you could see what you got working that would be great but I completely understand if you don't get the time.
